google/cloud/spanner_v1/testing/mock_spanner.py

```diff
@@ -51,23 +58,25 @@ def pop_error(self, context):
             context.abort_with_status(error)
 
     def get_result_as_partial_result_sets(
-        self, sql: str
+        self, sql: str, started_transaction: transaction.Transaction
     ) -> [result_set.PartialResultSet]:
         result: result_set.ResultSet = self.get_result(sql)
         partials = []
         first = True
         if len(result.rows) == 0:
             partial = result_set.PartialResultSet()
-            partial.metadata = result.metadata
+            partial.metadata = ResultSetMetadata(result.metadata)
             partials.append(partial)
         else:
             for row in result.rows:
                 partial = result_set.PartialResultSet()
                 if first:
-                    partial.metadata = result.metadata
+                    partial.metadata = ResultSetMetadata(result.metadata)
                 partial.values.extend(row)
                 partials.append(partial)
         partials[len(partials) - 1].stats = result.stats
+        if started_transaction:
+            partials[0].metadata.transaction = started_transaction
```

Comment:
Is there a specific reason we're including the transaction object details with the first partial result set?

Reply:
Yes, that is how the Spanner API works:
1. The first `PartialResultSet` in a stream should always contain the metadata for the query result. The remaining `PartialResultSet`s will not have this field set.
1. If the request (e.g. the `ExecuteSqlRequest`) contained a `BeginTransaction` option, then the metadata should also include the transaction ID of the transaction that was started as a side-effect of executing the request.

See https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.TransactionSelector for a bit more background information.